# TASK #1: UNDERSTAND THE PROBLEM STATEMENT AND BUSINESS CASE


<table>
  <tr><td>
    <img src="https://drive.google.com/uc?id=1yJKgmHrRFnBk987HJfeDrMcTEXtk0z7W"
         alt="Fashion MNIST sprite"  width="1000">
  </td></tr>
  <tr><td align="center">
    <b>Figure 1. Employee Retention Prediction
  </td></tr>
</table>


![alt text](https://drive.google.com/uc?id=10NJUOTWOBzp2MNkgcPpCF0fLtdoN_jKj)

![alt text](https://drive.google.com/uc?id=1evbDHoW2t0emxkbQd8yevYFZ5woJKRPY)

![alt text](https://drive.google.com/uc?id=1Mk2H7VYfv6ijUS9XqEdBQV6_LaHiyvkJ)

# TASK #2: IMPORT LIBRARIES AND DATASETS

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
# You will need to mount your drive using the following commands:
# For more information regarding mounting, please check this out: https://stackoverflow.com/questions/46986398/import-data-into-google-colaboratory

# from google.colab import drive
# drive.mount('/content/drive')

In [4]:
# You have to include the full link to the csv file containing your dataset
data = pd.read_csv('data.csv')
data.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [6]:
percent_missing = data.isnull().sum() * 100 / len(data)
# missing_value_df = pd.DataFrame({'column_name': data.columns,'percent_missing': percent_missing})
missing_value_df = pd.DataFrame({'percent_missing': percent_missing})
# print(percent_missing)
missing_value_df

percent_missing
Age                                   0.0
Attrition                             0.0
BusinessTravel                        0.0
DailyRate                             0.0
Department                            0.0
DistanceFromHome                      0.0
Education                             0.0
EducationField                        0.0
EmployeeCount                         0.0
EmployeeNumber                        0.0
EnvironmentSatisfaction               0.0
Gender                                0.0
HourlyRate                            0.0
JobInvolvement                        0.0
JobLevel                              0.0
JobRole                               0.0
JobSatisfaction                       0.0
MaritalStatus                         0.0
MonthlyIncome                         0.0
MonthlyRate                           0.0
NumCompaniesWorked                    0.0
Over18                                0.0
OverTime                              0.0
PercentSalaryHike                     0.0
PerformanceRating                     0.0
RelationshipSatisfaction              0.0
StandardHours                         0.0
StockOptionLevel                      0.0
TotalWorkingYears                     0.0
TrainingTimesLastYear                 0.0
WorkLifeBalance                       0.0
YearsAtCompany                        0.0
YearsInCurrentRole                    0.0
YearsSinceLastPromotion               0.0
YearsWithCurrManager                  0.0

In [7]:
data['Department'].value_counts()

Research & Development    961
Sales                     446
Human Resources            63
Name: Department, dtype: int64

In [8]:
data.describe()

Age    DailyRate  DistanceFromHome    Education  EmployeeCount  \
count  1470.000000  1470.000000       1470.000000  1470.000000         1470.0   
mean     36.923810   802.485714          9.192517     2.912925            1.0   
std       9.135373   403.509100          8.106864     1.024165            0.0   
min      18.000000   102.000000          1.000000     1.000000            1.0   
25%      30.000000   465.000000          2.000000     2.000000            1.0   
50%      36.000000   802.000000          7.000000     3.000000            1.0   
75%      43.000000  1157.000000         14.000000     4.000000            1.0   
max      60.000000  1499.000000         29.000000     5.000000            1.0   

       EmployeeNumber  EnvironmentSatisfaction   HourlyRate  JobInvolvement  \
count     1470.000000              1470.000000  1470.000000     1470.000000   
mean      1024.865306                 2.721769    65.891156        2.729932   
std        602.024335                 1.093082    20.329428        0.711561   
min          1.000000                 1.000000    30.000000        1.000000   
25%        491.250000                 2.000000    48.000000        2.000000   
50%       1020.500000                 3.000000    66.000000        3.000000   
75%       1555.750000                 4.000000    83.750000        3.000000   
max       2068.000000                 4.000000   100.000000        4.000000   

          JobLevel  JobSatisfaction  MonthlyIncome   MonthlyRate  \
count  1470.000000      1470.000000    1470.000000   1470.000000   
mean      2.063946         2.728571    6502.931293  14313.103401   
std       1.106940         1.102846    4707.956783   7117.786044   
min       1.000000         1.000000    1009.000000   2094.000000   
25%       1.000000         2.000000    2911.000000   8047.000000   
50%       2.000000         3.000000    4919.000000  14235.500000   
75%       3.000000         4.000000    8379.000000  20461.500000   
max       5.000000         4.000000   19999.000000  26999.000000   

       NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
count         1470.000000        1470.000000        1470.000000   
mean             2.693197          15.209524           3.153741   
std              2.498009           3.659938           0.360824   
min              0.000000          11.000000           3.000000   
25%              1.000000          12.000000           3.000000   
50%              2.000000          14.000000           3.000000   
75%              4.000000          18.000000           3.000000   
max              9.000000          25.000000           4.000000   

       RelationshipSatisfaction  StandardHours  StockOptionLevel  \
count               1470.000000         1470.0       1470.000000   
mean                   2.712245           80.0          0.793878   
std                    1.081209            0.0          0.852077   
min                    1.000000           80.0          0.000000   
25%                    2.000000           80.0          0.000000   
50%                    3.000000           80.0          1.000000   
75%                    4.000000           80.0          1.000000   
max                    4.000000           80.0          3.000000   

       TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
count        1470.000000            1470.000000      1470.000000   
mean           11.279592               2.799320         2.761224   
std             7.780782               1.289271         0.706476   
min             0.000000               0.000000         1.000000   
25%             6.000000               2.000000         2.000000   
50%            10.000000               3.000000         3.000000   
75%            15.000000               3.000000         3.000000   
max            40.000000               6.000000         4.000000   

       YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
count     1470.000000         1470.000000              1470.

# TASK #3: VISUALIZE DATASET

In [9]:
data['Attrition'].value_counts()

No     1233
Yes     237
Name: Attrition, dtype: int64

In [10]:
data['OverTime'].value_counts()

No     1054
Yes     416
Name: OverTime, dtype: int64

In [11]:
data['Over18'].value_counts()

Y    1470
Name: Over18, dtype: int64

In [12]:
from sklearn.preprocessing import LabelEncoder
# Let's replace 'Attritition' , 'overtime' , 'Over18' column with integers before performing any visualizations 
le_attrition = LabelEncoder()
data['Attrition'] = le_attrition.fit_transform(data['Attrition'].values)

le_overtime = LabelEncoder()
data['OverTime'] = le_overtime.fit_transform(data['OverTime'].values)

le_over18 = LabelEncoder()
data['Over18'] = le_over18.fit_transform(data['Over18'].values)

In [13]:
data.head()

Age  Attrition     BusinessTravel  DailyRate              Department  \
0   41          1      Travel_Rarely       1102                   Sales   
1   49          0  Travel_Frequently        279  Research & Development   
2   37          1      Travel_Rarely       1373  Research & Development   
3   33          0  Travel_Frequently       1392  Research & Development   
4   27          0      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked  Over18  OverTime  PercentSalaryHike  \
0        19479                   8       0         1                 11   
1        24907                   1       0         0                 23   
2         2396                   6       0         1                 15   
3        23159                   1       0         1                 11   
4        16632                   9       0         0                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [14]:
# Let's see if we have any missing data, luckily we don't!
percent_missing = data.isnull().sum() * 100 / len(data)
# missing_value_df = pd.DataFrame({'column_name': data.columns,'percent_missing': percent_missing})
missing_value_df = pd.DataFrame({'percent_missing': percent_missing})
# print(percent_missing)
missing_value_df

percent_missing
Age                                   0.0
Attrition                             0.0
BusinessTravel                        0.0
DailyRate                             0.0
Department                            0.0
DistanceFromHome                      0.0
Education                             0.0
EducationField                        0.0
EmployeeCount                         0.0
EmployeeNumber                        0.0
EnvironmentSatisfaction               0.0
Gender                                0.0
HourlyRate                            0.0
JobInvolvement                        0.0
JobLevel                              0.0
JobRole                               0.0
JobSatisfaction                       0.0
MaritalStatus                         0.0
MonthlyIncome                         0.0
MonthlyRate                           0.0
NumCompaniesWorked                    0.0
Over18                                0.0
OverTime                              0.0
PercentSalaryHike                     0.0
PerformanceRating                     0.0
RelationshipSatisfaction              0.0
StandardHours                         0.0
StockOptionLevel                      0.0
TotalWorkingYears                     0.0
TrainingTimesLastYear                 0.0
WorkLifeBalance                       0.0
YearsAtCompany                        0.0
YearsInCurrentRole                    0.0
YearsSinceLastPromotion               0.0
YearsWithCurrManager                  0.0

In [15]:
%matplotlib widget
sns.heatmap(data.isnull(), yticklabels=False, cbar=False)
plt.tight_layout()
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [16]:
data.hist(bins=30, figsize=(13,20), color='red', alpha=0.5)
plt.tight_layout()
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [17]:
# Several features such as 'MonthlyIncome' and 'TotalWorkingYears' are tail heavy
# It makes sense to drop 'EmployeeCount' and 'Standardhours' since they do not change from one employee to the other
data = data.drop(columns=['EmployeeNumber', 'EmployeeCount', 'StandardHours', 'Over18'])

In [18]:
# It makes sense to drop 'EmployeeCount' , 'Standardhours' and 'Over18' since they do not change from one employee to the other
# Let's drop 'EmployeeNumber' as well
data

Age  Attrition     BusinessTravel  DailyRate              Department  \
0      41          1      Travel_Rarely       1102                   Sales   
1      49          0  Travel_Frequently        279  Research & Development   
2      37          1      Travel_Rarely       1373  Research & Development   
3      33          0  Travel_Frequently       1392  Research & Development   
4      27          0      Travel_Rarely        591  Research & Development   
...   ...        ...                ...        ...                     ...   
1465   36          0  Travel_Frequently        884  Research & Development   
1466   39          0      Travel_Rarely        613  Research & Development   
1467   27          0      Travel_Rarely        155  Research & Development   
1468   49          0  Travel_Frequently       1023                   Sales   
1469   34          0      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                    1          2  Life Sciences                        2   
1                    8          1  Life Sciences                        3   
2                    2          2          Other                        4   
3                    3          4  Life Sciences                        4   
4                    2          1        Medical                        1   
...                ...        ...            ...                      ...   
1465                23          2        Medical                        3   
1466                 6          1        Medical                        4   
1467                 4          3  Life Sciences                        2   
1468                 2          3        Medical                        4   
1469                 8          3        Medical                        2   

      Gender  HourlyRate  JobInvolvement  JobLevel                    JobRole  \
0     Female          94               3         2            Sales Executive   
1       Male          61               2         2         Research Scientist   
2       Male          92               2         1      Laboratory Technician   
3     Female          56               3         1         Research Scientist   
4       Male          40               3         1      Laboratory Technician   
...      ...         ...             ...       ...                        ...   
1465    Male          41               4         2      Laboratory Technician   
1466    Male          42               2         3  Healthcare Representative   
1467    Male          87               4         2     Manufacturing Director   
1468    Male          63               2         2            Sales Executive   
1469    Male          82               4         2      Laboratory Technician   

      JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
0                   4        Single           5993        19479   
1                   2       Married           5130        24907   
2                   3        Single           2090         2396   
3                   3       Married           2909        23159   
4                   2       Married           3468        16632   
...               ...           ...            ...          ...   
1465                4       Married           2571        12290   
1466                1       Married           9991        21457   
1467                2       Married           6142         5174   
1468                2       Married           5390        13243   
1469                3       Married           4404        10228   

      NumCompaniesWorked  OverTime  PercentSalaryHike  PerformanceRating  \
0                      8         1                 11                  3   
1                      1         0                 23                  4   
2                      6         1                 15                  3   
3                      1         1                 11         

In [19]:
# Let's see how many employees left the company! 


In [20]:
# Count the number of employees who stayed and left
# It seems that we are dealing with an imbalanced dataset 
# employees that left and stayed the company
data_left = data.loc[data['Attrition'] == 1]
data_left.shape

(237, 31)

In [21]:
data_stayed = data.loc[data['Attrition'] == 0]
data_stayed.shape

(1233, 31)

In [22]:
%matplotlib widget
# Relative numbers
s_rel = data['Attrition'].value_counts(normalize=True) * 100.0
s_abs = data['Attrition'].value_counts()
fig, axs = plt.subplots(1, 2)
s_rel.plot(kind='bar', ax=axs[0], color=['darkorange', 'darkblue'])
axs[0].set_ylabel('Percent of total')

s_abs.plot(kind='bar', ax=axs[1], color=['darkorange', 'darkblue'])
axs[1].set_ylabel('Number')
plt.tight_layout()
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [23]:
data_stayed.describe()

Age  Attrition    DailyRate  DistanceFromHome    Education  \
count  1233.000000     1233.0  1233.000000       1233.000000  1233.000000   
mean     37.561233        0.0   812.504461          8.915653     2.927007   
std       8.888360        0.0   403.208379          8.012633     1.027002   
min      18.000000        0.0   102.000000          1.000000     1.000000   
25%      31.000000        0.0   477.000000          2.000000     2.000000   
50%      36.000000        0.0   817.000000          7.000000     3.000000   
75%      43.000000        0.0  1176.000000         13.000000     4.000000   
max      60.000000        0.0  1499.000000         29.000000     5.000000   

       EnvironmentSatisfaction   HourlyRate  JobInvolvement     JobLevel  \
count              1233.000000  1233.000000     1233.000000  1233.000000   
mean                  2.771290    65.952149        2.770479     2.145985   
std                   1.071132    20.380754        0.692050     1.117933   
min                   1.000000    30.000000        1.000000     1.000000   
25%                   2.000000    48.000000        2.000000     1.000000   
50%                   3.000000    66.000000        3.000000     2.000000   
75%                   4.000000    83.000000        3.000000     3.000000   
max                   4.000000   100.000000        4.000000     5.000000   

       JobSatisfaction  MonthlyIncome   MonthlyRate  NumCompaniesWorked  \
count      1233.000000    1233.000000   1233.000000          1233.00000   
mean          2.778589    6832.739659  14265.779400             2.64558   
std           1.093277    4818.208001   7102.260749             2.46009   
min           1.000000    1051.000000   2094.000000             0.00000   
25%           2.000000    3211.000000   7973.000000             1.00000   
50%           3.000000    5204.000000  14120.000000             2.00000   
75%           4.000000    8834.000000  20364.000000             4.00000   
max           4.000000   19999.000000  26997.000000             9.00000   

          OverTime  PercentSalaryHike  PerformanceRating  \
count  1233.000000        1233.000000        1233.000000   
mean      0.234388          15.231144           3.153285   
std       0.423787           3.639511           0.360408   
min       0.000000          11.000000           3.000000   
25%       0.000000          12.000000           3.000000   
50%       0.000000          14.000000           3.000000   
75%       0.000000          18.000000           3.000000   
max       1.000000          25.000000           4.000000   

       RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
count               1233.000000       1233.000000        1233.000000   
mean                   2.733982          0.845093          11.862936   
std                    1.071603          0.841985           7.760719   
min                    1.000000          0.000000           0.000000   
25%                    2.000000          0.000000           6.000000   
50%                    3.000000          1.000000          10.000000   
75%                    4.000000          1.000000          16.000000   
max                    4.000000          3.000000          38.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count            1233.000000      1233.000000     1233.000000   
mean                2.832928         2.781022        7.369019   
std                 1.293585         0.681907        6.096298   
min                 0.000000         1.000000        0.000000   
25%                 2.000000         2.000000        3.000000   
50%                 3.000000         3.000000        6.000000   
75%                 3.000000         3.000000       10.000000   
max                 6.000000         4.000000       37.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count         1233.000000              1233.000000           1233.000000  
mean             4.484185         

In [24]:
data_left.describe()

Age  Attrition    DailyRate  DistanceFromHome   Education  \
count  237.000000      237.0   237.000000        237.000000  237.000000   
mean    33.607595        1.0   750.362869         10.632911    2.839662   
std      9.689350        0.0   401.899519          8.452525    1.008244   
min     18.000000        1.0   103.000000          1.000000    1.000000   
25%     28.000000        1.0   408.000000          3.000000    2.000000   
50%     32.000000        1.0   699.000000          9.000000    3.000000   
75%     39.000000        1.0  1092.000000         17.000000    4.000000   
max     58.000000        1.0  1496.000000         29.000000    5.000000   

       EnvironmentSatisfaction  HourlyRate  JobInvolvement    JobLevel  \
count               237.000000  237.000000      237.000000  237.000000   
mean                  2.464135   65.573840        2.518987    1.637131   
std                   1.169791   20.099958        0.773405    0.940594   
min                   1.000000   31.000000        1.000000    1.000000   
25%                   1.000000   50.000000        2.000000    1.000000   
50%                   3.000000   66.000000        3.000000    1.000000   
75%                   4.000000   84.000000        3.000000    2.000000   
max                   4.000000  100.000000        4.000000    5.000000   

       JobSatisfaction  MonthlyIncome   MonthlyRate  NumCompaniesWorked  \
count       237.000000     237.000000    237.000000          237.000000   
mean          2.468354    4787.092827  14559.308017            2.940928   
std           1.118058    3640.210367   7208.153264            2.678519   
min           1.000000    1009.000000   2326.000000            0.000000   
25%           1.000000    2373.000000   8870.000000            1.000000   
50%           3.000000    3202.000000  14618.000000            1.000000   
75%           3.000000    5916.000000  21081.000000            5.000000   
max           4.000000   19859.000000  26999.000000            9.000000   

         OverTime  PercentSalaryHike  PerformanceRating  \
count  237.000000         237.000000         237.000000   
mean     0.535865          15.097046           3.156118   
std      0.499768           3.770294           0.363735   
min      0.000000          11.000000           3.000000   
25%      0.000000          12.000000           3.000000   
50%      1.000000          14.000000           3.000000   
75%      1.000000          17.000000           3.000000   
max      1.000000          25.000000           4.000000   

       RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
count                237.000000        237.000000         237.000000   
mean                   2.599156          0.527426           8.244726   
std                    1.125437          0.856361           7.169204   
min                    1.000000          0.000000           0.000000   
25%                    2.000000          0.000000           3.000000   
50%                    3.000000          0.000000           7.000000   
75%                    4.000000          1.000000          10.000000   
max                    4.000000          3.000000          40.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count             237.000000       237.000000      237.000000   
mean                2.624473         2.658228        5.130802   
std                 1.254784         0.816453        5.949984   
min                 0.000000         1.000000        0.000000   
25%                 2.000000         2.000000        1.000000   
50%                 2.000000         3.000000        3.000000   
75%                 3.000000         3.000000        7.000000   
max                 6.000000         4.000000       40.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count          237.000000               237.000000            237.000000  
mean             2.902954                 1.945148              2.852321  
std

In [25]:
data_stayed.describe() - data_left.describe()

Age  Attrition   DailyRate  DistanceFromHome   Education  \
count  996.000000      996.0  996.000000        996.000000  996.000000   
mean     3.953638       -1.0   62.141591         -1.717259    0.087345   
std     -0.800990        0.0    1.308860         -0.439892    0.018758   
min      0.000000       -1.0   -1.000000          0.000000    0.000000   
25%      3.000000       -1.0   69.000000         -1.000000    0.000000   
50%      4.000000       -1.0  118.000000         -2.000000    0.000000   
75%      4.000000       -1.0   84.000000         -4.000000    0.000000   
max      2.000000       -1.0    3.000000          0.000000    0.000000   

       EnvironmentSatisfaction  HourlyRate  JobInvolvement    JobLevel  \
count               996.000000  996.000000      996.000000  996.000000   
mean                  0.307155    0.378310        0.251491    0.508855   
std                  -0.098659    0.280797       -0.081355    0.177339   
min                   0.000000   -1.000000        0.000000    0.000000   
25%                   1.000000   -2.000000        0.000000    0.000000   
50%                   0.000000    0.000000        0.000000    1.000000   
75%                   0.000000   -1.000000        0.000000    1.000000   
max                   0.000000    0.000000        0.000000    0.000000   

       JobSatisfaction  MonthlyIncome  MonthlyRate  NumCompaniesWorked  \
count       996.000000     996.000000   996.000000          996.000000   
mean          0.310234    2045.646832  -293.528617           -0.295348   
std          -0.024781    1177.997634  -105.892515           -0.218428   
min           0.000000      42.000000  -232.000000            0.000000   
25%           1.000000     838.000000  -897.000000            0.000000   
50%           0.000000    2002.000000  -498.000000            1.000000   
75%           1.000000    2918.000000  -717.000000           -1.000000   
max           0.000000     140.000000    -2.000000            0.000000   

         OverTime  PercentSalaryHike  PerformanceRating  \
count  996.000000         996.000000         996.000000   
mean    -0.301477           0.134097          -0.002833   
std     -0.075980          -0.130783          -0.003328   
min      0.000000           0.000000           0.000000   
25%      0.000000           0.000000           0.000000   
50%     -1.000000           0.000000           0.000000   
75%     -1.000000           1.000000           0.000000   
max      0.000000           0.000000           0.000000   

       RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
count                996.000000        996.000000         996.000000   
mean                   0.134826          0.317667           3.618210   
std                   -0.053834         -0.014376           0.591515   
min                    0.000000          0.000000           0.000000   
25%                    0.000000          0.000000           3.000000   
50%                    0.000000          1.000000           3.000000   
75%                    0.000000          0.000000           6.000000   
max                    0.000000          0.000000          -2.000000   

       TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
count             996.000000       996.000000      996.000000   
mean                0.208455         0.122794        2.238217   
std                 0.038801        -0.134546        0.146314   
min                 0.000000         0.000000        0.000000   
25%                 0.000000         0.000000        2.000000   
50%                 1.000000         0.000000        3.000000   
75%                 0.000000         0.000000        3.000000   
max                 0.000000         0.000000       -3.000000   

       YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
count          996.000000               996.000000            996.000000  
mean             1.581231                 0.289240              1.515076  
std              0.47

In [26]:
#  Let's compare the mean and std of the employees who stayed and left 
# 'age': mean age of the employees who stayed is higher compared to who left
# 'DailyRate': Rate of employees who stayed is higher
# 'DistanceFromHome': Employees who stayed live closer to home 
# 'EnvironmentSatisfaction' & 'JobSatisfaction': Employees who stayed are generally more satisifed with their jobs
# 'StockOptionLevel': Employees who stayed tend to have higher stock option level

In [27]:
%matplotlib widget
sns.set(font_scale=0.7) 
correlations = data.corr()
f, ax = plt.subplots(figsize=(12,12))
sns.heatmap(correlations, annot=True, cmap='viridis', fmt="0.2f")
plt.tight_layout()
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [28]:
# Job level is strongly correlated with total working hours
# Monthly income is strongly correlated with Job level
# Monthly income is strongly correlated with total working hours
# Age is stongly correlated with monthly income

In [29]:
%matplotlib widget
plt.figure(figsize=(10,8))
sns.countplot(x='Age', hue='Attrition', data=data)
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [36]:
%matplotlib widget
df_age = data.groupby('Age')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
f, ax = plt.subplots(figsize=(10,5))
sns.barplot(ax=ax, data=df_age, x='Age', y='percent', hue='Attrition')
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [72]:
%matplotlib widget
df_age = data.groupby('Age')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
df_age1 = df_age.pivot(columns='Attrition', index='Age')
f, ax = plt.subplots(figsize=(10,5))
df_age1.plot(kind='bar', ax=ax, stacked=True)
ax.legend(["Stayed", "Left"], loc='lower right')
plt.ylabel('Percent attrition')
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [77]:
plt.figure(figsize=(12,15))

plt.subplot(411)
sns.countplot(x='JobRole', hue='Attrition', data=data)

plt.subplot(412)
sns.countplot(x='MaritalStatus', hue='Attrition', data=data)

plt.subplot(413)
sns.countplot(x='JobInvolvement', hue='Attrition', data=data)

plt.subplot(414)
sns.countplot(x='JobLevel', hue='Attrition', data=data)

plt.tight_layout()
plt.show()


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [79]:
%matplotlib widget


plt.figure(figsize=(12,15))

plt.subplot(411)
ax = plt.gca()
df_0 = data.groupby('JobRole')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
df_0 = df_0.pivot(columns='Attrition', index='JobRole')
df_0.plot(kind='bar', ax=ax, stacked=True)
ax.legend(["Stayed", "Left"], loc='lower right')
ax.set_ylabel('Percent attrition by JobRole')

plt.subplot(412)
ax = plt.gca()
df_0 = data.groupby('MaritalStatus')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
df_0 = df_0.pivot(columns='Attrition', index='MaritalStatus')
df_0.plot(kind='bar', ax=ax, stacked=True)
ax.legend(["Stayed", "Left"], loc='lower right')
ax.set_ylabel('Percent attrition by MaritalStatus')

plt.subplot(413)
ax = plt.gca()
df_0 = data.groupby('JobInvolvement')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
df_0 = df_0.pivot(columns='Attrition', index='JobInvolvement')
df_0.plot(kind='bar', ax=ax, stacked=True)
ax.legend(["Stayed", "Left"], loc='lower right')
ax.set_ylabel('Percent attrition by JobInvolvement')


plt.subplot(414)
ax = plt.gca()
df_0 = data.groupby('JobLevel')['Attrition'].value_counts(normalize=True).mul(100).rename('percent').reset_index()
df_0 = df_0.pivot(columns='Attrition', index='JobLevel')
df_0.plot(kind='bar', ax=ax, stacked=True)
ax.legend(["Stayed", "Left"], loc='lower right')
ax.set_ylabel('Percent attrition by JobLevel')


plt.tight_layout()
plt.show()


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [ ]:
# Single employees tend to leave compared to married and divorced
# Sales Representitives tend to leave compared to any other job 
# Less involved employees tend to leave the company 
# Less experienced (low job level) tend to leave the company 

In [96]:
# KDE is used for visualizing the Probability Density of a continuous variable
%matplotlib widget
plt.figure()
sns.kdeplot(data_left['DistanceFromHome'].values, label='Left', shade=True, color='darkorange')
plt.hist(data_left['DistanceFromHome'].values, bins=30, color='darkorange', ls=':', density=True, histtype='step')
sns.kdeplot(data_stayed['DistanceFromHome'].values, label='Stayed', shade=True, color='darkblue')
plt.hist(data_stayed['DistanceFromHome'].values, bins=30, color='darkblue', ls=':', density=True, histtype='step')

plt.xlabel('Distance from home')
plt.ylabel('Estimated density')
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [129]:

# KDE is used for visualizing the Probability Density of a continuous variable
%matplotlib widget
plt.figure()

data_plot = data.copy()
data_plot['x'] = 1
sns.violinplot(x='x', y='DistanceFromHome', hue="Attrition", data=data_plot, inner='quartile', bw=0.2, palette="muted", split=True)
plt.xticks([])
plt.ylabel('Distance from home')
plt.xlabel(' ')


ax = plt.gca()
handles, labels = ax.get_legend_handles_labels()
ax.legend(handles, ['Stayed', 'Left'])
plt.show()


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [130]:
%matplotlib widget
data_plot = data.copy()
data_plot['x'] = 1

plt.figure(figsize=(6,8))

plt.subplot(211)
sns.violinplot(x='x', y='TotalWorkingYears', hue="Attrition", data=data_plot, inner='quartile', bw=0.2, palette="muted", split=True)
plt.xticks([])
plt.ylabel('TotalWorkingYears')
plt.xlabel(' ')

ax = plt.gca()
handles, labels = ax.get_legend_handles_labels()
ax.legend(handles, ['Stayed', 'Left'])

plt.subplot(212)
sns.violinplot(x='x', y='YearsWithCurrManager', hue="Attrition", data=data_plot, inner='quartile', bw=0.2, palette="muted", split=True)
plt.xticks([])
plt.ylabel('YearsWithCurrManager')
plt.xlabel(' ')
ax = plt.gca()
handles, labels = ax.get_legend_handles_labels()
ax.legend(handles, ['Stayed', 'Left'])

plt.tight_layout()
plt.show()



Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [144]:
# Let's see the Gender vs. Monthly Income
plt.figure()
sns.violinplot(y='MonthlyIncome', x='Gender', data=data, inner='quartile', bw=0.2, palette="Set2")
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [152]:
# Let's see the Job Role vs. Monthly Income
plt.figure(figsize=(12,6))
chart = sns.violinplot(y='MonthlyIncome', x='JobRole', data=data, inner='quartile', bw=0.2)
chart.set_xticklabels(chart.get_xticklabels(), rotation=45, horizontalalignment='right')
plt.tight_layout()
plt.show()

/Users/hkromer/anaconda3/envs/tf/lib/python3.7/site-packages/ipykernel_launcher.py:2: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).
  


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

# TASK #4: CREATE TESTING AND TRAINING DATASET & PERFORM DATA CLEANING

In [190]:
# categorical variables
cat_var = ['BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus']
X_cat = data[cat_var]

In [191]:
X_cat

BusinessTravel              Department EducationField  Gender  \
0         Travel_Rarely                   Sales  Life Sciences  Female   
1     Travel_Frequently  Research & Development  Life Sciences    Male   
2         Travel_Rarely  Research & Development          Other    Male   
3     Travel_Frequently  Research & Development  Life Sciences  Female   
4         Travel_Rarely  Research & Development        Medical    Male   
...                 ...                     ...            ...     ...   
1465  Travel_Frequently  Research & Development        Medical    Male   
1466      Travel_Rarely  Research & Development        Medical    Male   
1467      Travel_Rarely  Research & Development  Life Sciences    Male   
1468  Travel_Frequently                   Sales        Medical    Male   
1469      Travel_Rarely  Research & Development        Medical    Male   

                        JobRole MaritalStatus  
0               Sales Executive        Single  
1            Research Scientist       Married  
2         Laboratory Technician        Single  
3            Research Scientist       Married  
4         Laboratory Technician       Married  
...                         ...           ...  
1465      Laboratory Technician       Married  
1466  Healthcare Representative       Married  
1467     Manufacturing Director       Married  
1468            Sales Executive       Married  
1469      Laboratory Technician       Married  

[1470 rows x 6 columns]

In [192]:
from sklearn.preprocessing import OneHotEncoder
ohc = OneHotEncoder()

X_cat = ohc.fit_transform(X_cat).toarray()

In [193]:
X_cat = pd.DataFrame(X_cat, columns=ohc.get_feature_names())
X_cat

x0_Non-Travel  x0_Travel_Frequently  x0_Travel_Rarely  \
0               0.0                   0.0               1.0   
1               0.0                   1.0               0.0   
2               0.0                   0.0               1.0   
3               0.0                   1.0               0.0   
4               0.0                   0.0               1.0   
...             ...                   ...               ...   
1465            0.0                   1.0               0.0   
1466            0.0                   0.0               1.0   
1467            0.0                   0.0               1.0   
1468            0.0                   1.0               0.0   
1469            0.0                   0.0               1.0   

      x1_Human Resources  x1_Research & Development  x1_Sales  \
0                    0.0                        0.0       1.0   
1                    0.0                        1.0       0.0   
2                    0.0                        1.0       0.0   
3                    0.0                        1.0       0.0   
4                    0.0                        1.0       0.0   
...                  ...                        ...       ...   
1465                 0.0                        1.0       0.0   
1466                 0.0                        1.0       0.0   
1467                 0.0                        1.0       0.0   
1468                 0.0                        0.0       1.0   
1469                 0.0                        1.0       0.0   

      x2_Human Resources  x2_Life Sciences  x2_Marketing  x2_Medical  \
0                    0.0               1.0           0.0         0.0   
1                    0.0               1.0           0.0         0.0   
2                    0.0               0.0           0.0         0.0   
3                    0.0               1.0           0.0         0.0   
4                    0.0               0.0           0.0         1.0   
...                  ...               ...           ...         ...   
1465                 0.0               0.0           0.0         1.0   
1466                 0.0               0.0           0.0         1.0   
1467                 0.0               1.0           0.0         0.0   
1468                 0.0               0.0           0.0         1.0   
1469                 0.0               0.0           0.0         1.0   

      x2_Other  x2_Technical Degree  x3_Female  x3_Male  \
0          0.0                  0.0        1.0      0.0   
1          0.0                  0.0        0.0      1.0   
2          1.0                  0.0        0.0      1.0   
3          0.0                  0.0        1.0      0.0   
4          0.0                  0.0        0.0      1.0   
...        ...                  ...        ...      ...   
1465       0.0                  0.0        0.0      1.0   
1466       0.0                  0.0        0.0      1.0   
1467       0.0                  0.0        0.0      1.0   
1468       0.0                  0.0        0.0      1.0   
1469       0.0                  0.0        0.0      1.0   

      x4_Healthcare Representative  x4_Human Resources  \
0                              0.0                 0.0   
1                              0.0                 0.0   
2                              0.0                 0.0   
3                              0.0                 0.0   
4                              0.0                 0.0   
...                            ...                 ...   
1465                           0.0                 0.0   
1466                           1.0                 0.0   
1467                           0.0                 0.0   
1468                           0.0                 0.0   
1469                           0.0                 0.0   

      x4_Laboratory Technician  x4_Manager  x4_Manufacturing Director  \
0                          0.0         0.0                        0.0   
1                          0.0         0.0         

In [194]:
# numerical variables
num_var = [s for s in data.columns.tolist() if s not in cat_var]
X_num = data[num_var]
X_num = X_num.drop(columns=['Attrition'])
X_num

Age  DailyRate  DistanceFromHome  Education  EnvironmentSatisfaction  \
0      41       1102                 1          2                        2   
1      49        279                 8          1                        3   
2      37       1373                 2          2                        4   
3      33       1392                 3          4                        4   
4      27        591                 2          1                        1   
...   ...        ...               ...        ...                      ...   
1465   36        884                23          2                        3   
1466   39        613                 6          1                        4   
1467   27        155                 4          3                        2   
1468   49       1023                 2          3                        4   
1469   34        628                 8          3                        2   

      HourlyRate  JobInvolvement  JobLevel  JobSatisfaction  MonthlyIncome  \
0             94               3         2                4           5993   
1             61               2         2                2           5130   
2             92               2         1                3           2090   
3             56               3         1                3           2909   
4             40               3         1                2           3468   
...          ...             ...       ...              ...            ...   
1465          41               4         2                4           2571   
1466          42               2         3                1           9991   
1467          87               4         2                2           6142   
1468          63               2         2                2           5390   
1469          82               4         2                3           4404   

      MonthlyRate  NumCompaniesWorked  OverTime  PercentSalaryHike  \
0           19479                   8         1                 11   
1           24907                   1         0                 23   
2            2396                   6         1                 15   
3           23159                   1         1                 11   
4           16632                   9         0                 12   
...           ...                 ...       ...                ...   
1465        12290                   4         0                 17   
1466        21457                   4         0                 15   
1467         5174                   1         1                 20   
1468        13243                   2         0                 14   
1469        10228                   2         0                 12   

      PerformanceRating  RelationshipSatisfaction  StockOptionLevel  \
0                     3                         1                 0   
1                     4                         4                 1   
2                     3                         2                 0   
3                     3                         3                 0   
4                     3                         4                 1   
...                 ...                       ...               ...   
1465                  3                         3                 1   
1466                  3                         1                 1   
1467                  4                         2                 1   
1468                  3                         4                 0   
1469                  3                         1                 0   

      TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  \
0                     8                      0                1   
1                    10                      3                3   
2                     7                      3                3   
3                     8                      3                3   
4                     6                      3                3   
...                 ...             

In [206]:
X_all = pd.concat([X_cat, X_num], axis=1)
X_all

x0_Non-Travel  x0_Travel_Frequently  x0_Travel_Rarely  \
0               0.0                   0.0               1.0   
1               0.0                   1.0               0.0   
2               0.0                   0.0               1.0   
3               0.0                   1.0               0.0   
4               0.0                   0.0               1.0   
...             ...                   ...               ...   
1465            0.0                   1.0               0.0   
1466            0.0                   0.0               1.0   
1467            0.0                   0.0               1.0   
1468            0.0                   1.0               0.0   
1469            0.0                   0.0               1.0   

      x1_Human Resources  x1_Research & Development  x1_Sales  \
0                    0.0                        0.0       1.0   
1                    0.0                        1.0       0.0   
2                    0.0                        1.0       0.0   
3                    0.0                        1.0       0.0   
4                    0.0                        1.0       0.0   
...                  ...                        ...       ...   
1465                 0.0                        1.0       0.0   
1466                 0.0                        1.0       0.0   
1467                 0.0                        1.0       0.0   
1468                 0.0                        0.0       1.0   
1469                 0.0                        1.0       0.0   

      x2_Human Resources  x2_Life Sciences  x2_Marketing  x2_Medical  \
0                    0.0               1.0           0.0         0.0   
1                    0.0               1.0           0.0         0.0   
2                    0.0               0.0           0.0         0.0   
3                    0.0               1.0           0.0         0.0   
4                    0.0               0.0           0.0         1.0   
...                  ...               ...           ...         ...   
1465                 0.0               0.0           0.0         1.0   
1466                 0.0               0.0           0.0         1.0   
1467                 0.0               1.0           0.0         0.0   
1468                 0.0               0.0           0.0         1.0   
1469                 0.0               0.0           0.0         1.0   

      x2_Other  x2_Technical Degree  x3_Female  x3_Male  \
0          0.0                  0.0        1.0      0.0   
1          0.0                  0.0        0.0      1.0   
2          1.0                  0.0        0.0      1.0   
3          0.0                  0.0        1.0      0.0   
4          0.0                  0.0        0.0      1.0   
...        ...                  ...        ...      ...   
1465       0.0                  0.0        0.0      1.0   
1466       0.0                  0.0        0.0      1.0   
1467       0.0                  0.0        0.0      1.0   
1468       0.0                  0.0        0.0      1.0   
1469       0.0                  0.0        0.0      1.0   

      x4_Healthcare Representative  x4_Human Resources  \
0                              0.0                 0.0   
1                              0.0                 0.0   
2                              0.0                 0.0   
3                              0.0                 0.0   
4                              0.0                 0.0   
...                            ...                 ...   
1465                           0.0                 0.0   
1466                           1.0                 0.0   
1467                           0.0                 0.0   
1468                           0.0                 0.0   
1469                           0.0                 0.0   

      x4_Laboratory Technician  x4_Manager  x4_Manufacturing Director  \
0                          0.0         0.0                        0.0   
1                          0.0         0.0         

In [207]:
# note that we dropped the target 'Atrittion'


In [208]:
from sklearn.preprocessing import MinMaxScaler
sc = MinMaxScaler()

X = sc.fit_transform(X_all)

In [209]:
X

array([[0.        , 0.        , 1.        , ..., 0.22222222, 0.        ,
        0.29411765],
       [0.        , 1.        , 0.        , ..., 0.38888889, 0.06666667,
        0.41176471],
       [0.        , 0.        , 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 1.        , ..., 0.11111111, 0.        ,
        0.17647059],
       [0.        , 1.        , 0.        , ..., 0.33333333, 0.        ,
        0.47058824],
       [0.        , 0.        , 1.        , ..., 0.16666667, 0.06666667,
        0.11764706]])

In [210]:
y = data['Attrition'].values

# TASK #5: UNDERSTAND THE INTUITION BEHIND LOGISTIC REGRESSION CLASSIFIERS, ARTIFICIAL NEURAL NETWORKS, AND RANDOM FOREST CLASSIFIER

![alt text](https://drive.google.com/uc?id=19DpnhFkfsNEDPlH1dkfdr1zO36vRcBit)

![alt text](https://drive.google.com/uc?id=1J03xZf6OiYtGV3IgJBUURBWyScpvaAbU)

![alt text](https://drive.google.com/uc?id=1WNsznVn7je5r9HGnSLLdABICxrIv2Mrs)

![alt text](https://drive.google.com/uc?id=1bX5uGmy5vbYTlp7m4tw_V2kTNzAHpHLp)

![alt text](https://drive.google.com/uc?id=1ztrMNehNYWMw6NwhOOC9BDBdnoNirpqZ)

# TASK #6: UNDERSTAND HOW TO ASSESS CLASSIFICATION MODELS 

![alt text](https://drive.google.com/uc?id=1OZLbKm1AJSyvoBgfvlfcLIWZxLOvzOWq)

![alt text](https://drive.google.com/uc?id=11pNdVw4oWeNOWrkadrrxon7FU4qO5m6U)

![alt text](https://drive.google.com/uc?id=1Bk1xFW2tGBdwg-njOhw79MxtYBQnK-6x)

![alt text](https://drive.google.com/uc?id=19cXoBqSiqbEGNofnD603bz3xEAsX28hy)

# TASK #7: TRAIN AND EVALUATE A LOGISTIC REGRESSION CLASSIFIER

In [247]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [248]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.linear_model import LogisticRegression

# specify the norm used in the penalization
penalty = ['l1', 'l2', 'elasticnet', 'none']
# Tolerance for stopping criteria
tol = [1e-8, 1e-6, 1e-4, 1e-2]
# Inverse of regularization strength
C = [1000, 100, 10, 1, 0.01, 0.001, 0.0001]
# Solver
solver = ['lbfgs', 'sag', 'liblinear']

# Create the random grid
random_grid = {'penalty': penalty,
               'tol': tol,
               'C': C,
               'solver': solver}


# Use the random grid to search for best hyperparameters
# First create the base model to tune
clf = LogisticRegression()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
clf_random = RandomizedSearchCV(estimator=clf, param_distributions=random_grid, n_iter=100, cv=3, verbose=2, random_state=42, n_jobs=-1)
# Fit the random search model
clf_random.fit(X_train, y_train)

Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  18 tasks      | elapsed:    0.2s
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:    2.7s finished


RandomizedSearchCV(cv=3, estimator=LogisticRegression(), n_iter=100, n_jobs=-1,
                   param_distributions={'C': [1000, 100, 10, 1, 0.01, 0.001,
                                              0.0001],
                                        'penalty': ['l1', 'l2', 'elasticnet',
                                                    'none'],
                                        'solver': ['lbfgs', 'sag', 'liblinear'],
                                        'tol': [1e-08, 1e-06, 0.0001, 0.01]},
                   random_state=42, verbose=2)

In [249]:
clf_random.best_params_

{'tol': 1e-06, 'solver': 'liblinear', 'penalty': 'l1', 'C': 1}

In [250]:
from sklearn.model_selection import GridSearchCV


# specify the norm used in the penalization
penalty = ['l2', 'elasticnet']
# Tolerance for stopping criteria
tol = [1e-6, 1e-4]
# Inverse of regularization strength
C = [30, 1, 0.03, 0.001]
# Solver
solver = ['lbfgs']

# Create the parameter grid based on the results of random search 
param_grid = {'penalty': penalty,
               'tol': tol,
               'C': C,
               'solver': solver}
# Create a based model
clf = LogisticRegression()
# Instantiate the grid search model
grid_search = GridSearchCV(estimator=clf, param_grid=param_grid, 
                          cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)
grid_search.best_params_

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.


Fitting 3 folds for each of 16 candidates, totalling 48 fits


[Parallel(n_jobs=-1)]: Done  48 out of  48 | elapsed:    0.2s finished


{'C': 1, 'penalty': 'l2', 'solver': 'lbfgs', 'tol': 1e-06}

In [255]:
from sklearn.metrics import classification_report
def evaluate(model, X_test, y_test):
    y_pred = model.predict(X_test)
    print('Model Performance')
    print(classification_report(y_test, y_pred))

#### Base model

In [256]:
base_model = LogisticRegression()
base_model.fit(X_train, y_train)
evaluate(base_model, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.91      0.97      0.94       320
           1       0.68      0.40      0.50        48

    accuracy                           0.90       368
   macro avg       0.80      0.68      0.72       368
weighted avg       0.88      0.90      0.88       368



#### Best RandomSearch

In [257]:
best_random = clf_random.best_estimator_
best_random.fit(X_train, y_train)
evaluate(best_random, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.91      0.97      0.94       320
           1       0.63      0.35      0.45        48

    accuracy                           0.89       368
   macro avg       0.77      0.66      0.70       368
weighted avg       0.87      0.89      0.87       368



#### Best GridSearch

In [258]:
best_grid = grid_search.best_estimator_
best_grid.fit(X_train, y_train)
evaluate(best_grid, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.91      0.97      0.94       320
           1       0.68      0.40      0.50        48

    accuracy                           0.90       368
   macro avg       0.80      0.68      0.72       368
weighted avg       0.88      0.90      0.88       368



In [259]:
# Testing Set Performance


# TASK #8: TRAIN AND EVALUATE A RANDOM FOREST CLASSIFIER

In [261]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [262]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier

# Number of trees in random forest
n_estimators = [int(x) for x in np.linspace(start = 100, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 150, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]

# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}


# Use the random grid to search for best hyperparameters
# First create the base model to tune
clf = RandomForestClassifier()
# Random search of parameters, using 3 fold cross validation, 
# search across 100 different combinations, and use all available cores
clf_random = RandomizedSearchCV(estimator=clf, param_distributions=random_grid, n_iter=100, cv=3, verbose=2, random_state=42, n_jobs=-1)
# Fit the random search model
clf_random.fit(X_train, y_train)

Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  17 tasks      | elapsed:    7.2s
[Parallel(n_jobs=-1)]: Done 138 tasks      | elapsed:   47.4s
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed:  2.0min finished


RandomizedSearchCV(cv=3, estimator=RandomForestClassifier(), n_iter=100,
                   n_jobs=-1,
                   param_distributions={'bootstrap': [True, False],
                                        'max_depth': [10, 24, 38, 52, 66, 80,
                                                      94, 108, 122, 136, 150,
                                                      None],
                                        'max_features': ['auto', 'sqrt'],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': [100, 311, 522, 733,
                                                         944, 1155, 1366, 1577,
                                                         1788, 2000]},
                   random_state=42, verbose=2)

In [263]:
clf_random.best_params_

{'n_estimators': 100,
 'min_samples_split': 5,
 'min_samples_leaf': 2,
 'max_features': 'auto',
 'max_depth': 122,
 'bootstrap': False}

In [264]:
from sklearn.model_selection import GridSearchCV


# Number of trees in random forest
n_estimators = [100, 200, 300, 400, 500]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [10, 150, 500]
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [False]

# Create the random grid
param_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

# Create a based model
clf = RandomForestClassifier()
# Instantiate the grid search model
grid_search = GridSearchCV(estimator=clf, param_grid=param_grid, 
                          cv=3, n_jobs=-1, verbose=2)
grid_search.fit(X_train, y_train)


Fitting 3 folds for each of 270 candidates, totalling 810 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 12 concurrent workers.
[Parallel(n_jobs=-1)]: Done  17 tasks      | elapsed:    2.5s
[Parallel(n_jobs=-1)]: Done 138 tasks      | elapsed:   17.2s
[Parallel(n_jobs=-1)]: Done 341 tasks      | elapsed:   40.4s
[Parallel(n_jobs=-1)]: Done 624 tasks      | elapsed:  1.3min
[Parallel(n_jobs=-1)]: Done 810 out of 810 | elapsed:  1.6min finished


GridSearchCV(cv=3, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'bootstrap': [False], 'max_depth': [10, 150, 500],
                         'max_features': ['auto', 'sqrt'],
                         'min_samples_leaf': [1, 2, 4],
                         'min_samples_split': [2, 5, 10],
                         'n_estimators': [100, 200, 300, 400, 500]},
             verbose=2)

In [265]:
grid_search.best_params_

{'bootstrap': False,
 'max_depth': 10,
 'max_features': 'auto',
 'min_samples_leaf': 1,
 'min_samples_split': 10,
 'n_estimators': 100}

In [266]:
from sklearn.metrics import classification_report
def evaluate(model, X_test, y_test):
    y_pred = model.predict(X_test)
    print('Model Performance')
    print(classification_report(y_test, y_pred))

#### Base model

In [267]:
base_model = RandomForestClassifier()
base_model.fit(X_train, y_train)
evaluate(base_model, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.88      0.99      0.93       320
           1       0.57      0.08      0.15        48

    accuracy                           0.87       368
   macro avg       0.72      0.54      0.54       368
weighted avg       0.84      0.87      0.83       368



#### Best RandomSearch

In [268]:
best_random = clf_random.best_estimator_
best_random.fit(X_train, y_train)
evaluate(best_random, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.88      0.99      0.93       320
           1       0.56      0.10      0.18        48

    accuracy                           0.87       368
   macro avg       0.72      0.55      0.55       368
weighted avg       0.84      0.87      0.83       368



#### Best GridSearch

In [269]:
best_grid = grid_search.best_estimator_
best_grid.fit(X_train, y_train)
evaluate(best_grid, X_test, y_test)

Model Performance
              precision    recall  f1-score   support

           0       0.88      0.99      0.93       320
           1       0.62      0.10      0.18        48

    accuracy                           0.88       368
   macro avg       0.75      0.55      0.56       368
weighted avg       0.85      0.88      0.83       368



# TASK #9: TRAIN AND EVALUATE A DEEP LEARNING MODEL 

In [270]:
import tensorflow as tf

In [273]:
X.shape

(1470, 50)

In [295]:
model = tf.keras.models.Sequential()

model.add(tf.keras.layers.Dense(units=500, activation='relu', input_shape=(50, )))
model.add(tf.keras.layers.Dense(units=500, activation='relu'))
model.add(tf.keras.layers.Dense(units=500, activation='relu'))
model.add(tf.keras.layers.Dense(units=500, activation='relu'))
model.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 500)               25500     
_________________________________________________________________
dense_7 (Dense)              (None, 500)               250500    
_________________________________________________________________
dense_8 (Dense)              (None, 500)               250500    
_________________________________________________________________
dense_9 (Dense)              (None, 500)               250500    
_________________________________________________________________
dense_10 (Dense)             (None, 1)                 501       
Total params: 777,501
Trainable params: 777,501
Non-trainable params: 0
_________________________________________________________________


In [296]:
model.compile(optimizer='Adam', loss='binary_crossentropy', metrics=['accuracy'])

In [297]:
history = model.fit(X_train, y_train, epochs=100, batch_size=50)

Train on 1102 samples
Epoch 1/100
1102/1102 [==============================] - 1s 844us/sample - loss: 0.4536 - accuracy: 0.8276
Epoch 2/100
1102/1102 [==============================] - 0s 173us/sample - loss: 0.3518 - accuracy: 0.8503
Epoch 3/100
1102/1102 [==============================] - 0s 168us/sample - loss: 0.3289 - accuracy: 0.8730
Epoch 4/100
1102/1102 [==============================] - 0s 170us/sample - loss: 0.2922 - accuracy: 0.8884
Epoch 5/100
1102/1102 [==============================] - 0s 166us/sample - loss: 0.2529 - accuracy: 0.9047
Epoch 6/100
1102/1102 [==============================] - 0s 164us/sample - loss: 0.2500 - accuracy: 0.9065
Epoch 7/100
1102/1102 [==============================] - 0s 164us/sample - loss: 0.2232 - accuracy: 0.9065
Epoch 8/100
1102/1102 [==============================] - 0s 164us/sample - loss: 0.1781 - accuracy: 0.9365
Epoch 9/100
1102/1102 [==============================] - 0s 166us/sample - loss: 0.1386 - accuracy: 0.9446
Epoch 10/100
11

In [300]:
%matplotlib widget
# summarize history for accuracy
plt.plot(history.history['accuracy'])
plt.title('model accuracy')
plt.ylabel('accuracy')
plt.xlabel('epoch')
plt.legend(['train', 'test'], loc='upper left')
plt.show()


Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [301]:
%matplotlib widget
# summarize history for loss
plt.plot(history.history['loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'test'], loc='upper left')
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [302]:
y_pred = model.predict(X_test)
y_pred = (y_pred > 0.5)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.95      0.93       320
           1       0.54      0.42      0.47        48

    accuracy                           0.88       368
   macro avg       0.73      0.68      0.70       368
weighted avg       0.87      0.88      0.87       368



# Oversampling


In [303]:
model_smote = tf.keras.models.Sequential()

model_smote.add(tf.keras.layers.Dense(units=500, activation='relu', input_shape=(50, )))
model_smote.add(tf.keras.layers.Dense(units=500, activation='relu'))
model_smote.add(tf.keras.layers.Dense(units=500, activation='relu'))
model_smote.add(tf.keras.layers.Dense(units=500, activation='relu'))
model_smote.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))
model_smote.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_11 (Dense)             (None, 500)               25500     
_________________________________________________________________
dense_12 (Dense)             (None, 500)               250500    
_________________________________________________________________
dense_13 (Dense)             (None, 500)               250500    
_________________________________________________________________
dense_14 (Dense)             (None, 500)               250500    
_________________________________________________________________
dense_15 (Dense)             (None, 1)                 501       
Total params: 777,501
Trainable params: 777,501
Non-trainable params: 0
_________________________________________________________________


In [304]:
from imblearn.over_sampling import SMOTE 
oversampler = SMOTE(random_state=0)
smote_train, smote_target = oversampler.fit_sample(X_train, y_train)
history_smote = model.fit(smote_train, smote_target, epochs = 100, batch_size = 50)


Train on 1826 samples
Epoch 1/100
1826/1826 [==============================] - 0s 200us/sample - loss: 0.4593 - accuracy: 0.8729
Epoch 2/100
1826/1826 [==============================] - 0s 182us/sample - loss: 0.1486 - accuracy: 0.9359
Epoch 3/100
1826/1826 [==============================] - 0s 184us/sample - loss: 0.0493 - accuracy: 0.9841
Epoch 4/100
1826/1826 [==============================] - 0s 181us/sample - loss: 0.0606 - accuracy: 0.9797
Epoch 5/100
1826/1826 [==============================] - 0s 181us/sample - loss: 0.0542 - accuracy: 0.9841
Epoch 6/100
1826/1826 [==============================] - 0s 185us/sample - loss: 0.0186 - accuracy: 0.9923
Epoch 7/100
1826/1826 [==============================] - 0s 183us/sample - loss: 0.0263 - accuracy: 0.9912
Epoch 8/100
1826/1826 [==============================] - 0s 185us/sample - loss: 0.0049 - accuracy: 0.9995
Epoch 9/100
1826/1826 [==============================] - 0s 192us/sample - loss: 0.0043 - accuracy: 0.9978
Epoch 10/100
18

In [305]:
%matplotlib widget
# summarize history for loss
plt.plot(history_smote.history['loss'])
plt.title('model loss')
plt.ylabel('loss')
plt.xlabel('epoch')
plt.legend(['train', 'test'], loc='upper left')
plt.show()

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [306]:
y_pred = model.predict(X_test)
y_pred = (y_pred > 0.5)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.92      0.92       320
           1       0.43      0.40      0.41        48

    accuracy                           0.85       368
   macro avg       0.67      0.66      0.66       368
weighted avg       0.85      0.85      0.85       368



precision    recall  f1-score   support

           0       0.92      0.95      0.93       320
           1       0.54      0.42      0.47        48

    accuracy                           0.88       368
   macro avg       0.73      0.68      0.70       368
weighted avg       0.87      0.88      0.87       368




In [ ]:
# Testing Set Performance



# EXCELLENT JOB! YOU SHOULD BE PROUD OF YOUR NEWLY ACQUIRED SKILLS